Draft Transport WG Blog Post (Roadmap) (#1985)
modelcontextprotocol/modelcontextprotocol: blog/content/posts/2025-12-xx-transport-direction.md

---
date: "2025-12-09T09:00:00+00:00"
publishDate: "2025-12-09T09:00:00+00:00"
title: "Transport Direction"
author: "Kurtis van Gent, Shaun Smith"
tags: ["mcp", "governance"]
ShowToc: true
---

> # MCP Transport Roadmap

Reviewer: ```suggestion
# MCP Transport Roadmap For Next Spec Iteration
```

> When MCP launched in November 2024, the most common way to use it was locally - connecting Clients to Servers over STDIO. As MCP has grown to become the standard choice for LLM integration, the need to deploy MCP in distributed environments - and at scale has grown.

Reviewer: ```suggestion
When MCP first launched in November of 2024, most users ran it locally, connecting clients to servers over STDIO. But as MCP has become the go-to standard for LLM integration, the community's needs have evolved. There's growing demand for distributed deployments that can operate at scale.
```

Pioneers of remote, scaled deployments using the Streamable HTTP transport have found practical challenges which make reusing existing infrastructure deployment patterns hard.

## Roadmap

> To address these challenges, the Transport Working Group has spent the last few months collecting feedback and ideas from the community. We have worked closely with the Core Maintainers to propose solutions to the most common problems.

Reviewer: ```suggestion
Over the past few months, the Transport Working Group has worked together with the community and MCP Core Maintainers to develop solutions to these challenges.
```

In this post we share the roadmap for evolving the Streamable HTTP transport, and invite community engagement to shape the future of the protocol.

### A Stateless Protocol

> At the outset, MCP was designed to be _stateful_, with Clients and Servers maintaining knowlege of each other over a bidirectional communication channel.

Reviewer 2: ```suggestion
At the outset, MCP was designed to be a _stateful_ protocol, with Clients and Servers maintaining knowledge of each other over a bidirectional communication channel.
```

> At the outset, MCP was designed to be _stateful_, with Clients and Servers maintaining knowlege of each other over a bidirectional communication channel.

Reviewer: ```suggestion
MCP was designed from the start as a stateful protocol, with clients and servers maintaining mutual awareness through a persistent, bidirectional channel.
```

Connections are initialized with a handshake which shares information like Capabilities and Protocol Version. This state remains fixed for the duration of the connection, and requires techniques such as sticky sessions or distributed session storage to for scaled deployment.

> We propose making MCP stateless by:

Reviewer 2: Instead of "we propose" maybe say "We intend to make MCP stateless by"

- Replacing the `initialize` handshake - and sending the shared information with each request/response instead.

> - Providing a `discovery` mechanism for Clients to query Server Capabilities if they need the information early (for example for UX purposes).

Reviewer: ```suggestion
- Providing a `discovery` mechanism for clients to query server capabilities if they need the information early, for scenarios such as UX hydration.
```

> We're moving toward a more dynamic model where clients can optimistically "just try" what they need and receive helpful error messages if unsupported.

Reviewer: ```suggestion
With these changes, we're moving toward a more dynamic model where clients can optimistically "just try" what they need and receive clear error messages if a capability is unsupported.
```

> > [!NOTE]

Reviewer: @evalstate @kurtisvg would recommend building this with Hugo locally - I am not sure if the GitHub note shortcode will render on the blog itself.

>
> Many SDKs present a _`stateless`_ option in their Server transport configuration. This flag actually controls whether the `Mcp-Session-Id` header is used - read below for more on sessions.

### Elevating Sessions

> Today, Sessions are a side effect of a transport connection. For STDIO, sessions are implicit in the process lifecycle. For Streamable HTTP, sessions are created when a Server assigns an `Mcp-Session-Id` during `intialize`. This mixes transport and application layer concerns.

Reviewer: ```suggestion
Today, sessions are a side effect of the transport connection. With STDIO, sessions are implicit in the process lifecycle. With Streamable HTTP, sessions are created when a server assigns an `Mcp-Session-Id` during initialization. This conflates transport and application layer concerns.
```

> We plan to move Sessions to the _data model layer_ - making them explicit rather than implicit.

Reviewer: ```suggestion
We plan to move sessions to the _data model layer_ - making them explicit rather than implicit.
```

> This means that MCP applications will be able to handle sessions as part of their domain logic. A cookie style mechanism is the preferred choice.

Reviewer 2: ```suggestion
This means that MCP applications will be able to handle sessions as part of their domain logic. We are currently investigating a way of accumulating state in the request using a cookie style mechanism as the preferred choice.
```

Author: Updated text.

> This makes MCP similar to standard HTTP - where the protocol itself is stateless, while applications build stateful semantics with cookies, tokens and similar mechanisms. The exact approach to session creation is still being worked on - but this removes existing ambiguities on what a session is in remote MCP.

Reviewer: ```suggestion
Following the approach above makes MCP very similar to standard HTTP, where the protocol itself is stateless while applications build stateful semantics with cookies, tokens, and similar mechanisms. The exact approach to session creation is still being designed. Our goal is to remove existing ambiguities on what a session is in remote MCP scenarios.
```

### Elicitations and Sampling

Two standout MCP features enable advanced AI workflows: [Elicitations](https://spec.modelcontextprotocol.io/specification/2025-03-26/client/elicitation) for requesting human input, and [Sampling](https://spec.modelcontextprotocol.io/specification/2025-03-26/client/sampling) for agentic LLM interactions.

In our stateless design, this bidirectional request pattern requires rethinking. Currently when a Server needs more information to complete a Tool Call, it suspends operation and waits for a Client response - meaning it must remember all of its outstanding requests.

> To avoid the need to do this, we'll make the Server Request/Response similar to the way Chat APIs work. This means that the Client will return both the Request _and_ Response allowing the Server to reconstruct the necessary state purely from the returned message.

Reviewer: ```suggestion
To avoid this problem, we'll make the server request and response similar to the way chat APIs work. The client will return both the request _and_ response, allowing the server to reconstruct the necessary state purely from the returned message.
```

> To avoid the need to do this, we'll make the Server Request/Response similar to the way Chat APIs work. This means that the Client will return both the Request _and_ Response allowing the Server to reconstruct the necessary state purely from the returned message.

Reviewer 2: ```suggestion
To avoid the need to do this, we'll make the Server Request/Response similar to how long running operation's work in many APIs. This means that a Server will be able to return an 'intermediate' result, which declares the need for more information to continue. The client will be able to send another request, this time with the elicitation or sampling request included. We believe this, along with the cookies style state, will allow many servers to handle these operations at scale and without necessarily the need for backend storage. 
```

> ### Update Notifications, Subscriptions

Reviewer: ```suggestion
### Update Notifications and Subscriptions
```

MCP's dynamic nature means that Tools, Prompts, and Resources can change during operation. The current protocol uses server-to-client `ListChangedNotification` messages as an optimization to prompt cache invalidation.

We're replacing the general-purpose GET stream with explicit subscription streams. Clients will start specialized streams when they want to subscribe to specific items, and can manage multiple concurrent subscriptions. If a subscription stream is interrupted, the client simply restarts it—no complex resumption logic required.

> To make notifications truly optional optimizations rather than requirements, we're adding TTLs and version numbers (potentially using etags) to data. This allows clients to make intelligent caching decisions without depending on the notification stream, improving reliability in lossy network conditions.

Reviewer: ```suggestion
To make notifications truly optional optimizations rather than requirements, we're adding Time-To-Live (TTLs) and version numbers (such as [ETag](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/ETag)) to data. This allows clients to make intelligent caching decisions without depending on the notification stream, improving reliability in lossy network conditions.
```

### JSON-RPC Envelopes

The protocol currently uses JSON-RPC for all message envelopes, including method names and parameters. As we optimize for HTTP deployments, questions arise about whether to move toward more traditional REST patterns.

While we don't yet have consensus on fully replacing JSON-RPC with REST, we agree that copying the JSON-RPC method name into the HTTP path (or a dedicated header) improves clarity and enables better HTTP caching semantics.

> The core tension is between protocol consistency (keeping JSON-RPC everywhere) and HTTP integration (making MCP look like normal web traffic). We're evaluating this tradeoff with real deployment scenarios in mind.

Reviewer 2: IIUC the discussion correctly, I think the argument is less about protocol consistency (between stdio and http) and more about consistency between new versions and olds ones (minimizing the work of still-evolving infra to keep up with MCP)

## Summary

These changes fundamentally reorient MCP around stateless, independent requests while preserving the rich features that make it powerful. For Server developers, this elimination of session state makes horizontal scaling trivial—no more sticky sessions or distributed session stores. For Client hosts, architecture becomes simpler and more predictable.

Most developers using SDKs will see minimal impact - and many will require no code changes at all. The primary difference is architectural: deployments become simpler, serverless platforms become viable for rich MCP features, and the protocol aligns better with modern infrastructure patterns.

## Next Steps

> We'll complete detailed designs in Q1 2026, targeting launch in H2 2026. Throughout this process, we invite community feedback and participation. These changes enable MCP to scale from local development to global deployments while maintaining the ergonomics that made it successful.

Reviewer 2: ```suggestion
We intend to being this work immediately, targeting in-spec changes for Q1 2026, and making them available in the next spec release (tentatively June 2026). Throughout this process, we invite community feedback and participation. These changes enable MCP to scale from local development to global deployments while maintaining the ergonomics that made it successful.
```

---
date: "2025-12-09T09:00:00+00:00"
publishDate: "2025-12-09T09:00:00+00:00"
title: "Evolving MCP Transports To Scale For Production"
author: "Kurtis Van Gent (Transport WG Maintainer), Shaun Smith (Transport WG Maintainer)"
tags: ["mcp", "governance", "transports"]
ShowToc: true
---

# Evolving MCP Transports To Scale For Production

When MCP first launched in November of 2024, most users ran it locally, connecting clients to servers over STDIO. But as MCP has become the go-to standard for LLM integration, the community's needs have evolved. There's growing demand for distributed deployments that can operate at scale.

> Early adopters of remote, scaled deployments using Streamable HTTP transport have encountered several practical challenges that made it difficult to leverage their existing infrastructure patterns.

Reviewer 2: nit: 
```suggestion
Early adopters of remote, scaled deployments using Streamable HTTP transport have encountered several practical challenges that made it difficult to leverage their existing infrastructure patterns. As we see enterprise MCP deployments scaling to millions of daily requests, the friction of stateful connections has become a bottleneck for managed services and load balancing.
```

We intend to make MCP stateless by:

- Replacing the `initialize` handshake and sending the shared information with each request/response instead.
- Providing a `discovery` mechanism for clients to query server capabilities if they need the information early, for scenarios such as UX hydration.

With these changes, we're moving toward a more dynamic model where clients can optimistically "just try" what they need and receive clear error messages if a capability is unsupported.

> **NOTE:** Many SDKs present a _`stateless`_ option in their Server transport configuration. This flag actually controls whether the `Mcp-Session-Id` header is used - read below for more on sessions.

### Elevating Sessions

Today, sessions are a side effect of the transport connection. With STDIO, sessions are implicit in the process lifecycle. With Streamable HTTP, sessions are created when a server assigns an `Mcp-Session-Id` during initialization. This conflates transport and application layer concerns.

We plan to move sessions to the _data model layer_ - making them explicit rather than implicit.

> This means that MCP applications will be able to handle sessions as part of their domain logic. We are investigating the right way to do this - with a cookie style mechanism being the preferred choice for handling this state.

Reviewer 2: ```suggestion
This means that MCP applications will be able to handle sessions as part of their domain logic. We are currently exploring options for this, with a cookie-like mechanism being the leading candidate to decouple session state from the transport layer.
```

Following the approach above makes MCP very similar to standard HTTP, where the protocol itself is stateless while applications build stateful semantics with cookies, tokens, and similar mechanisms. The exact approach to session creation is still being designed. Our goal is to remove existing ambiguities on what a session is in remote MCP scenarios.

### Elicitations and Sampling

Two existing MCP features enable key AI workflows that developers want: [Elicitations](https://spec.modelcontextprotocol.io/specification/2025-11-25/client/elicitation) for requesting human input, and [Sampling](https://spec.modelcontextprotocol.io/specification/2025-11-25/client/sampling) for agentic LLM interactions.

To support these features at scale, the bidirectional communication pattern that these features rely on requires rethinking. Currently when a server needs more information to complete a tool call, it suspends operation and waits for a client response - meaning that it must remember all of its outstanding requests.

> To avoid this problem, we'll make the server request and response similar to the way chat APIs work. The Server will return the Elicitation request as normal - however the client will return both request _and_ response, allowing the server to reconstruct the necessary state purely from the returned message. This avoids managing potentially long-running state between specific nodes, and potentially removing the need for back-end storage.

Reviewer 2: nit: 
```suggestion
To avoid this problem, we'll make the server request and response similar to the way chat APIs work. The Server will return the Elicitation request as normal - however the client will return both request _and_ response, allowing the server to reconstruct the necessary state purely from the returned message. This architectural shift eliminates the need for expensive 'sticky' load balancers or shared storage backends (like Redis) for simple ephemeral tool calls. Developers can build complex, multi-turn agentic behaviors on serverless platforms without necessarily managing persistent infrastructure.
```
